DataStructures/BinarySearchTree/BinarySearchTree.cs

```diff
@@ -0,0 +1,410 @@
+        private BinarySearchTreeNode<TKey> GetMin(BinarySearchTreeNode<TKey> node)
+        {
+            if (node.Left == null)
+            {
+                return node;
+            }
+
+            return GetMin(node.Left);
+        }
+
+        /// <summary>
+        /// Recursive method to get a list with the keys sorted in in-order order.
+        /// </summary>
+        /// <param name="node">Node to traverse from.</param>
+        /// <returns>List of keys in in-order order.</returns>
+        private IList<TKey> GetKeysInOrder(BinarySearchTreeNode<TKey>? node)
+        {
+            if (node == null)
+            {
+                return new List<TKey>();
+            }
+
+            var result = new List<TKey>();
+            result.AddRange(GetKeysInOrder(node.Left));
+            result.Add(node.Key);
+            result.AddRange(GetKeysInOrder(node.Right));
+            return result;
+        }
+
+        /// <summary>
+        /// Recursive method to get a list with the keys sorted in pre-order order.
+        /// </summary>
+        /// <param name="node">Node to traverse from.</param>
+        /// <returns>List of keys in pre-order order.</returns>
+        private IList<TKey> GetKeysPreOrder(BinarySearchTreeNode<TKey>? node)
+        {
+            if (node == null)
+            {
+                return new List<TKey>();
+            }
+
+            var result = new List<TKey>();
+            result.Add(node.Key);
+            result.AddRange(GetKeysPreOrder(node.Left));
+            result.AddRange(GetKeysPreOrder(node.Right));
+            return result;
+        }
+
+        /// <summary>
+        /// Recursive method to get a list with the keys sorted in post-order order.
+        /// </summary>
+        /// <param name="node">Node to traverse from.</param>
+        /// <returns>List of keys in post-order order.</returns>
+        private IList<TKey> GetKeysPostOrder(BinarySearchTreeNode<TKey>? node)
+        {
+            if (node == null)
+            {
+                return new List<TKey>();
+            }
+
+            var result = new List<TKey>();
+            result.AddRange(GetKeysPostOrder(node.Left));
+            result.AddRange(GetKeysPostOrder(node.Right));
+            result.Add(node.Key);
+            return result;
+        }
+
+        /// <summary>
+        /// Recursive method to find a node with a matching key.
+        /// </summary>
+        /// <param name="node">Node to search from.</param>
+        /// <param name="key">Key to find.</param>
+        /// <returns>The node with the specified if it exists, a default value otherwise.</returns>
+        private BinarySearchTreeNode<TKey>? Search(BinarySearchTreeNode<TKey>? node, TKey key)
+        {
+            if (node == null)
+            {
+                return default;
+            }
+
+            int compareResult = node.Key.CompareTo(key);
+            if (compareResult > 0)
```

Comment:
Codacy found an issue: [Use the '?:' operator here.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6277689)